hash-tables/overview-of-hash-tables.md

```diff
@@ -0,0 +1,153 @@
+# Overview of Hash Tables
+
+## Learning Goals
+
+- Define hash table
+- Define hash function
+- Describe how hash tables resolve collisions
+
+## Introduction
+
+Python dictionaries are fascinating! They store key-value pairs, and, as long as we have the key, getting the value has a time complexity of O(1).
+
+This is in direct comparison to arrays/lists. We could store elements in an array, but finding specific entries in an array requires looping through potentially the entire list an O(n) operation.
```

Comment:
develop a comparison between lists and dicts
my main motivation for this was that the existing text mixes the notions of quickly looking up a key verses finding a value.
arrays _do_ have O(1) key lookups, but their keys are restricted to integers. But even in a dict, if we are truly looking for a _value_ in the dict, we would still need to iterate. They _keys_ can be looked up in O(1).
I know this is what was intended, but I wanted to make sure this is an apples to apples example.
```suggestion
Let's say we had a list of names `["Mickey", "Kermit", "Garfield"]`. If we let a user enter a name as input, and wanted to check whether that name was in this list, we would need to iterate over the array and check whether the user input matched each name. This is a O(n) operation.

Note that using the `in` operator on lists only helps the readability, not the performance. Internally, it still needs to check whether the value we are looking for matches any of the values in the array, one by one!

But what if we used a dictionary? Let's store our list of names as the keys of a dictionary like `{"Mickey": 1, "Kermit": 1, "Garfield": 1}`. The values don't really matter, but are required since dictionaries store key-value pairs. Here, `1` works just as well as any other value we might pick.

By storing our "list" in this way, we can now use the `in` operator on dictionaries to ask whether the user input is a key in our dictionary in a time complexity of O(1), constant time!
```